### PR TITLE
Upstream serialization improvements

### PR DESCRIPTION
**Pull request by str4d**
Cherry-picked from the following upstream PRs:

- bitcoin/bitcoin#5264
- bitcoin/bitcoin#6914
- bitcoin/bitcoin#6215
- bitcoin/bitcoin#8068
  - Only the `COMPACTSIZE` wrapper commit
- bitcoin/bitcoin#8658
- bitcoin/bitcoin#8708
  - Only the serializer variadics commit
- bitcoin/bitcoin#9039
- bitcoin/bitcoin#9125
  - Only the first two commits (the last two block on other upstream PRs)

Part of #2074.

### CODE REVIEW

**Comment by str4d** [comment]
@zkbot try

**Comment by str4d** [comment]
Transient RPC test failure in one builder; all the other supported builders passed.

**Comment by ebfull** [comment]
ACK

File: src/primitives/transaction.cpp [hunk 1]
```diff
@@ -123,7 +123,7 @@ std::string CTxIn::ToString() const
     if (prevout.IsNull())
         str += strprintf(", coinbase %s", HexStr(scriptSig));
     else
-        str += strprintf(", scriptSig=%s", scriptSig.ToString().substr(0,24));
+        str += strprintf(", scriptSig=%s", HexStr(scriptSig).substr(0, 24));
```
Comment (arielgabizon) [inline src/primitives/transaction.cpp]: wondering if there was a reason HexStr wasn't used here before

File: src/gtest/test_foundersreward.cpp [hunk 2]
```diff
@@ -106,11 +106,11 @@ TEST(founders_reward_test, general) {
     // address = t2ENg7hHVqqs9JwU5cgjvSbxnT2a9USNfhy
     // script.ToString() = OP_HASH160 55d64928e69829d9376c776550b6cc710d427153 OP_EQUAL
     // HexStr(script) = a91455d64928e69829d9376c776550b6cc710d42715387
-    EXPECT_EQ(params.GetFoundersRewardScriptAtHeight(1), ParseHex("a914ef775f1f997f122a062fff1a2d7443abd1f9c64287"));
+    EXPECT_EQ(HexStr(params.GetFoundersRewardScriptAtHeight(1)), "a914ef775f1f997f122a062fff1a2d7443abd1f9c64287");
```
Comment (arielgabizon) [inline src/gtest/test_foundersreward.cpp]: what does this have to do with the prevector?

File: src/primitives/transaction.cpp (hunk 1 above)
Comment (arielgabizon) [reply src/primitives/transaction.cpp]: Asked about it on bitcoin, seems fine.https://github.com/bitcoin/bitcoin/pull/5264#discussion_r182417714

**Comment by bitcartel** [comment]
@zkbot try

**Comment by bitcartel** [comment]
@zkbot retry

File: src/gtest/test_foundersreward.cpp (hunk 2 above)
Comment (bitcartel) [reply src/gtest/test_foundersreward.cpp]: Good catch, the change to this file should be dropped.

Comment (ebfull) [reply src/gtest/test_foundersreward.cpp]: There is no implementation of equality now after the other changes, so this change is necessary or it won't compile.

**Comment by ebfull** [comment]
@zkbot r+

File: src/gtest/test_foundersreward.cpp (hunk 2 above)
Comment (arielgabizon) [reply src/gtest/test_foundersreward.cpp]: I don't see an equality operator for CScript being erased in this commit or the previous one.
So does that mean some commits in this PR don't compile?

Comment (arielgabizon) [reply src/gtest/test_foundersreward.cpp]: I guess what's going on is when there's a ToString method it can compare as strings and that was erased in the prevector commit. So the third commit doesn't compile, and the fourth does https://github.com/zcash/zcash/pull/3180/commits/d2fb34fb7ca56e612bfc619224aa27cd914aeb55
I guess that's fine.

File: src/serialize.h [hunk 3]
```diff
@@ -1146,4 +1020,16 @@ inline void SerReadWriteMany(Stream& s, int nType, int nVersion, CSerActionUnser
     ::UnserializeMany(s, nType, nVersion, args...);
 }
 
+template <typename T>
+size_t GetSerializeSize(const T& t, int nType, int nVersion = 0)
```
Comment (arielgabizon) [inline src/serialize.h]: Is there a reason to have the default value of 0 for `nVersion`?
Seems worth thinking of especially with all our transaction versions (though I don't remember if we are using `nVersion` for those)

Comment (arielgabizon) [reply src/serialize.h]: A side benefit of reviewing this is that I finally understood what that weird `ADD_SERIALIZE_METHODS` macro is!

File: src/chain.h [hunk 4]
```diff
@@ -380,7 +380,7 @@ class CDiskBlockIndex : public CBlockIndex
 
         // Only read/write nSproutValue if the client version used to create
         // this index was storing them.
-        if ((nType & SER_DISK) && (nVersion >= SPROUT_VALUE_VERSION)) {
```
Comment (arielgabizon) [inline src/chain.h]: the previous commit has errors `nType` not defined here and doesn't compile.
This change might be moved to the previous commit

Comment (arielgabizon) [reply src/chain.h]: Ahh OK, this line wasn't in the bitcoin code, that's why it's still wrong here.

File: src/serialize.h (hunk 3 above)
Comment (str4d) [reply src/serialize.h]: `GetSerializeSize` is not just for transactions; it is used by *all* serialization operations in `zcashd`. The justification for leaving the default as 0 is that changing it could likely break a bunch of other parsers.

File: src/gtest/test_foundersreward.cpp (hunk 2 above)
Comment (str4d) [reply src/gtest/test_foundersreward.cpp]: It's not that it was erased, just that upstream didn't implement it. The change here is definitely a no-op, as comparing two hex strings should always be the same as comparing their decoded forms.

File: src/serialize.h (hunk 3 above)
Comment (arielgabizon) [reply src/serialize.h]: Ahh, yeah now I see a default of 0 in all the old versions of the method.

**Comment by arielgabizon** [comment]
postmerge ack

File: src/gtest/test_foundersreward.cpp (hunk 2 above)
Comment (daira) [reply src/gtest/test_foundersreward.cpp]: Provided the case matches, and ``HexStr`` always produces lowercase so that should be fine.

File: src/serialize.h [hunk 5]
```diff
@@ -1098,4 +1099,52 @@ class CSizeComputer
...
+}
+
+template<typename Stream, typename Arg>
+void SerializeMany(Stream& s, int nType, int nVersion, Arg&& arg)
+{
+    ::Serialize(s, std::forward<Arg>(arg), nType, nVersion);
+}
+
+template<typename Stream, typename Arg, typename... Args>
+void SerializeMany(Stream& s, int nType, int nVersion, Arg&& arg, Args&&... args)
```
Comment (daira) [inline src/serialize.h]: What's that klunking sound I hear? Is it the sound of language features being duplicated?

File: src/chain.h [hunk 6]
```diff
@@ -340,8 +340,9 @@ class CDiskBlockIndex : public CBlockIndex
     ADD_SERIALIZE_METHODS;
 
     template <typename Stream, typename Operation>
-    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        if (!(nType & SER_GETHASH))
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        int nVersion = s.GetVersion();
+        if (!(s.GetType() & SER_GETHASH))
```
Comment (daira) [inline src/chain.h]: Why is the version excluded when computing the hash? Don't we want the hash to change if the version does?

(I know this is not affected by this commit.)

File: src/primitives/block.h [hunk 7]
```diff
@@ -185,8 +185,9 @@ struct CBlockLocator
     ADD_SERIALIZE_METHODS;
 
     template <typename Stream, typename Operation>
-    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        if (!(nType & SER_GETHASH))
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        int nVersion = s.GetVersion();
+        if (!(s.GetType() & SER_GETHASH))
```
Comment (daira) [inline src/primitives/block.h]: Why is the version excluded when computing the hash? Don't we want the hash to change if the version does?

(I know this is not affected by this commit.)

File: src/chain.h [hunk 8]
```diff
@@ -380,7 +380,7 @@ class CDiskBlockIndex : public CBlockIndex
 
         // Only read/write nSproutValue if the client version used to create
         // this index was storing them.
-        if ((nType & SER_DISK) && (nVersion >= SPROUT_VALUE_VERSION)) {
+        if ((s.GetType() & SER_DISK) && (nVersion >= SPROUT_VALUE_VERSION)) {
```
Comment (daira) [inline src/chain.h]: ``nVersion`` here refers to ``s.GetVersion()``, which may be different from the block header version ``this->nVersion``. I suggest replacing ``nVersion`` with ``s.GetVersion()`` here and on line 346 to make it more explicit.

Comment (daira) [reply src/chain.h]: That is, although the local doesn't *technically* shadow the member variable, there is a way to refer to the member variable (as just ``nVersion``) that would usually be correct and here is not. That's confusing.

**Comment by daira** [comment]
GitHub is not showing one of my review comments because it was touched by later code, but it is still relevant: https://github.com/zcash/zcash/pull/3180#discussion_r183018835

File: src/key.h [hunk 9]
```diff
@@ -170,11 +170,28 @@ struct CExtKey {
...
+        unsigned int len = BIP32_EXTKEY_SIZE;
+        ::WriteCompactSize(s, len);
+        unsigned char code[BIP32_EXTKEY_SIZE];
+        Encode(code);
+        s.write((const char *)&code[0], len);
+    }
+    template <typename Stream>
+    void Unserialize(Stream& s)
+    {
+        unsigned int len = ::ReadCompactSize(s);
```
Comment (daira) [inline src/key.h]: This differs from the near-duplicate code [in pubkey.h](https://github.com/zcash/zcash/pull/3180/files#diff-10ed0089ccea825883e83e5afcef7697R231) by not checking ``len``. Why is there near-duplicate code and why does it differ?
